cognite/pygen/utils/external_id_factories.py

```diff
@@ -37,116 +66,193 @@ def windmill_factory(domain_cls: type, data: dict) -> str:
+    This creates a sha256 hash based on the data.
 
+    Args:
+        domain_cls: The domain class
+        data: The data for the domain class instance
 
-def create_sha256_factory(shorten: bool = True) -> Callable[[type, dict], str]:
+    Returns:
+        A sha256 hash
     """
-    This creates a sha256 hash external id factory.
+    return sha256_(str(data).encode()).hexdigest()
+
+
+INCREMENTAL_ID_REGISTRY: dict[str, int] = defaultdict(int)
+
+
+def incremental_id(domain_cls: type, data: dict) -> str:
+    """
+    This creates an incremental id for each domain class instance.
+    WARNING: This is not recommended for production use due to potential conflicts in threads and external services
+    potentially using the same incremental indexing.
 
     Args:
-        shorten: If True, the external id will be shortened to 7 characters.
+        domain_cls: The domain class
+        data: The data for the domain class instance
 
     Returns:
-        A factory function that can be set on the DomainModelApply class.
+        An incremental integer given the domain class
     """
+    global INCREMENTAL_ID_REGISTRY
+    domain_cls_name = domain_cls.__name__
+    INCREMENTAL_ID_REGISTRY[domain_cls_name] += 1
 
-    def sha256_factory_wrapped(domain_cls: type, data: dict) -> str:
-        """
-        This creates a sha256 hash external id for each domain class.
+    return str(INCREMENTAL_ID_REGISTRY[domain_cls_name])
 
-        Args:
-            domain_cls: The domain class
-            data: The data to create the external id from
 
-        Returns:
-            A sha256 hash of the data
+class ExternalIdFactory:
+    """
+    This is a factory class that can be used to create external ids for domain class instances.
+    """
+
+    def __init__(self, factory: Callable[[type, dict], str], shorten_length: Optional[int] = None):
+        self.factory = factory
 
-        """
-        prefix = domain_cls.__name__.casefold().removesuffix("apply")
-        hash_ = sha256(str(data).encode()).hexdigest()
-        if shorten:
-            return f"{prefix}:{hash_[:7]}"
+        if shorten_length is not None and shorten_length <= 0:
+            raise ValueError("Shorten length must be greater than 0 or None")
         else:
-            return f"{prefix}:{hash_}"
+            self.shorten_length = shorten_length or 7
+
+    def __call__(self, domain_cls: type, data: dict) -> str:
+        return self.factory(domain_cls, data)
+
+    def short(self, domain_cls: type, data: dict) -> str:
+        return shorten_string(self.factory(domain_cls, data), self.shorten_length)
+
+
+def domain_name_factory(shorten_length=None):
+    return ExternalIdFactory(factory=domain_name, shorten_length=shorten_length)
+
+
+def uuid_factory(shorten_length=None):
+    return ExternalIdFactory(factory=uuid, shorten_length=shorten_length)
+
+
+def sha256_factory(shorten_length=None):
+    return ExternalIdFactory(factory=sha256, shorten_length=shorten_length)
```

Comment:
Surprised that this passed `mypy`, shouldn't this be?
```suggestion
def domain_name_factory(shorten_length=None) -> ExternalIdFactory:
    return ExternalIdFactory(factory=domain_name, shorten_length=shorten_length)


def uuid_factory(shorten_length=None) -> ExternalIdFactory:
    return ExternalIdFactory(factory=uuid, shorten_length=shorten_length)


def sha256_factory(shorten_length=None) -> ExternalIdFactory:
    return ExternalIdFactory(factory=sha256, shorten_length=shorten_length)
```

Reply:
`mypy` default behavior does not check anything that isn't typed because it interprets it as `Any`. I can make another PR to configure `disallow_untyped_defs` so these things would get picked up